src/hotspot/share/opto/split_if.cpp

```diff
@@ -220,6 +221,29 @@ bool PhaseIdealLoop::split_up( Node *n, Node *blk1, Node *blk2 ) {
     }
   }
 
+  if (n->Opcode() == Op_Opaque1) {
+    CountedLoopNode* cl = n->as_Opaque1()->guarded_counted_loop();
+    // If this Opaque1 is part of the zero trip guard for a loop:
+    // 1- it can't be shared
+    // 2- the zero trip guard can't be the if that's being split
+    // As a consequence, this node could be assigned control anywhere between its current control and the zero trip guard.
+    // Move it down to get it out of the way of split if and avoid breaking the zero trip guard shape.
+    if (cl != NULL) {
+      // Find the zero trip guard from the loop
+      Node* ctrl = cl->skip_predicates();
+      assert(ctrl->in(0)->is_If(), "expect and If from the zero trip guard");
+      assert(ctrl->in(0) != blk1 && ctrl->in(0) != blk2, "shouldn't split a zero trip guard");
+      Node* bol = ctrl->in(0)->in(1);
+      assert(bol->is_Bool(), "bad zero trip guard shape");
+      Node* cmp = bol->in(1);
+      assert(cmp->Opcode() == Op_CmpI, "bad zero trip guard shape");
+      set_ctrl(n, ctrl->in(0)->in(0));
+      set_ctrl(cmp, ctrl->in(0)->in(0));
+      set_ctrl(bol, ctrl->in(0)->in(0));
```

Comment:
Why you assign control to `cmp` and `bol` too?

Reply:
The subgraph is` (Bool (CmpI (Opaque1...))`. So if only the control of `Opaque1` is updated then the `CmpI`/`Bool` could end up with a control that strictly dominates the control of the `Opaque1`. It's quite possible that that wouldn't break anything but wouldn't it be inconsistent and quite ugly?

Reply:
I mean we never assign control edge to `Bool` and `Cmp` nodes - they depend only on their inputs. At least I don't know about it.

Reply:
`set_ctrl()` doesn't change the control input of the nodes, right? It only updates the current loop opts pass's table of controls and all data nodes are in that table. I'm confused by what could be wrong here.

Reply:
You are right, I forgot that this function only update site table. Good.